File: scripts/core/entities/maintenance.py
from pydantic import BaseModel, Field
from enum import Enum
from typing import List, Optional
from datetime import datetime

class FailureRiskLevel(str, Enum):
    low = "low"
    medium = "medium"
    high = "high"

class FailurePrediction(BaseModel):
    riskLevel: FailureRiskLevel = Field(..., description="Risk level of failure (low, medium, high)")
    estimatedFailureDate: Optional[datetime] = Field(None, description="Predicted failure date if available")
    reason: str = Field(..., description="Reason for failure prediction")

class Maintenance(BaseModel):
    id: str
    maintenanceId: str = Field(..., description="The unique ID of the maintenance")
    componentId: str = Field(..., description="The ID of the component affected")
    subcomponentsAffected: List[str] = Field([], description="List of affected subcomponent IDs")
    date: datetime = Field(..., description="Date of maintenance")
    performedBy: str = Field(..., description="User ID who performed the maintenance")
    cost: float = Field(..., description="Cost of the maintenance")
    description: Optional[str] = Field(..., description="Description of the maintenance performed")
    nextMaintenance: Optional[datetime] = Field(None, description="Scheduled next maintenance date")
    failurePrediction: Optional[FailurePrediction] = Field(None, description="Prediction of potential future failures")

File: web/src/domain/entities/maintenance.ts
import { z } from "zod";

export const FailureRiskLevel = z.enum(["low", "medium", "high"]);
export type FailureRiskLevel = z.infer<typeof FailureRiskLevel>;

export const FailurePredictionSchema = z.object({
  riskLevel: FailureRiskLevel.describe(
    "Risk level of failure (low, medium, high)",
  ),
  estimatedFailureDate: z.coerce
    .date()
    .nullable()
    .describe("Predicted failure date if available"),
  reason: z.string().describe("Reason for failure prediction"),
});
export type FailurePrediction = z.infer<typeof FailurePredictionSchema>;

export const MaintenanceSchema = z.object({
  id: z.string(),
  maintenanceId: z.string().describe("The unique ID of the maintenance"),
  componentId: z.string().describe("The ID of the component affected"),
  subcomponentsAffected: z
    .array(z.string())
    .default([])
    .describe("List of affected subcomponent IDs"),
  date: z.coerce.date().describe("Date of maintenance"),
  performedBy: z.string().describe("User ID who performed the maintenance"),
  cost: z.number().describe("Cost of the maintenance"),
  description: z
    .string()
    .nullable()
    .describe("Description of the maintenance performed"),
  nextMaintenance: z.coerce
    .date()
    .nullable()
    .describe("Scheduled next maintenance date"),
  failurePrediction: FailurePredictionSchema.nullable().describe(
    "Prediction of potential future failures",
  ),
});

export type Maintenance = z.infer<typeof MaintenanceSchema>;
